Pal 880 update root documents for streaming (#315)
gchq/Palisade-services: README.md

# <img src="logos/logo.svg" width="180">

> ### A Tool for Complex and Scalable Data Access Policy Enforcement

Reviewer: ```suggestion
## A Tool for Complex and Scalable Data Access Policy Enforcement
```

## Overview

From the client’s perspective, they submit a request to examine data and receive a reference for this data. 
This reference can then be used to view the data after it has been retrieved and possibly redacted or filtered based on the context of the query and the permissions of the user making the request.

> Under the hood, the request involves sending a message that is processed sequentially by a set of the micro-services starting first with Palisade Service.

Reviewer 2: A note after this saying something like "This service returns a token, the reference to be used for the rest of the request"

Author: Have tied the idea of a reference and a token in the description.

With each service, the message is modified and then forwarded onto the next service with the end goal of producing resources ready to be viewed by the client.

> This initial sequence of steps is completed with the Filtered-Resource Service prepared to provide the processed data.

Reviewer 2: Another note, "This service returns, one-by-one upon request over a websocket, metadata for a single resource (e.g. a file) they are authorised to view in some capacity."

Author: Had a go at weaving this into the sentence "In the subsequent request by the client ...."  .

> In the subsequent request by the client to the Data Service, the Filtered-Resource is requested to then proceed with passing this data to the client.

Reviewer 2: This doesn't read too good, maybe something like
```suggestion
In the subsequent request by the client to the Data Service, the service reads the resource and returns it to the client (after applying relevant policies to the data).
```
?

> The following diagram shows in a high level the services and their relative relationships with each other. 

Reviewer 2: ```suggestion
The following diagram shows at a high level the services and their relative relationships with each other. 
```

Communications between the services and their related support servers utilise Kafka streaming. 

For a detailed description of the services' workflow see [Appendix I - Palisade Flow Diagram](#appendix-i).  
For a more technical description of each of the services, follow the links below.

<!--- 
See doc/services.drawio for the source of this diagram
--->

> ![Service Sequence diagram](doc/services.png)

Reviewer 2: Data-Svc goes to AMS, not FRS
TOS goes from AMS to FRS as current, but AMS also goes straight to FRS (like a triangle shape between the three)
Audit Svc lines could be tidier

Author: Ok had another go at it.

> [Attribute-Masking Service](attribute-masking-service/README.md) creates a redacted version of the rescue.  

Reviewer: redacted version of the rescue?

> [Audit Service](audit-service/README.md) operates in the background providing an audit record of the process.  

Reviewer: ```suggestion
[Audit Service](audit-service/README.md) used for auditing successful and failed requests for data in Palisade.  
```

> [Data Service](data-service/README.md) initiates the process of passing the resources to the client.  

Reviewer: doesnt sound accurate. the Data Service is used to connect to actual backing stores containing data to returned processed resources

> [Filtered-Resource Service](filtered-resource-service/README.md) readies the resources and later handles passing the data to the client.  

Reviewer: again doesnt sound totally accurate

> [Palisade Service](palisade-service/README.md) takes client resource requests and starts the process.  

Reviewer: ```suggestion
[Palisade Service](palisade-service/README.md) recieves a clients request, returns a unique token and initiates the processing of resources.
```

> [Policy Service](policy-service/README.md) provides the policy that are to be applied to the data.  

Reviewer: ```suggestion
[Policy Service](policy-service/README.md) provides the policies that apply to the resources.
```

> [Resource Service](resource-service/README.md) identifies the resources that have been requested.  

Reviewer: ```suggestion
[Resource Service](resource-service/README.md) communicates with the backing stores to ensure the resource exists.
```

> [Topic-Offset Service](topic-offset-service/README.md) identifies the start and end of the resources for a request.  

Reviewer: something like this sounds better
```
[Topic-Offset Service](topic-offset-service/README.md) listens to kafka topics to inform the Filtered-Resource Service on when messages are ready to be processed.
```

Reviewer 2: ```suggestion
[Topic-Offset Service](topic-offset-service/README.md) listens to kafka topics to inform the Filtered-Resource Service where on the topic (at what offset) the first message for each token appears.
```

> [User Service](user-service/README.md) identifies the user in the context of the Palisade service.  

Reviewer 2: Order these as they process messages, not alphabetically:
* palisade
* user
* resource
* attribute-masking
* topic-offset
* filtered-resource

_leave a gap_

* data

_leave a gap_

* audit

Author: done

> [User Service](user-service/README.md) identifies the user in the context of the Palisade service.  

Reviewer: ```suggestion
[User Service](user-service/README.md) connects to User Service providers to ensure the user exists.  
```

# <img src="logos/logo.svg" width="180">

### A Tool for Complex and Scalable Data Access Policy Enforcement

## Overview

From the client’s perspective, they submit a request to examine data and receive a reference for this data. 

> This reference can then be used to view the data after it has been retrieved and possibly redacted or filtered based on the context of the query and the permissions for the user making the request.

Reviewer: ```suggestion
This reference can then be used to view the data after it has been retrieved and possibly redacted or filtered based on the context of the query and the permissions of the user making the request.
```

Under the hood, the request involves sending a message that is processed sequentially by a set of the micro-services starting first with Palisade Service.
With each service, the message is modified and then forwarded onto the next service with the end goal of producing resources ready to be viewed by the client.
This initial sequence of steps is completed with the Filtered-Resource Service prepared to provide the processed data.
In the subsequent request by the client to the Data Service, the Filtered-Resource is requested to then proceed with passing this data to the client.
The following diagram shows in a high level the services and their relative relationships with each other. 
Communications between the services and their related support servers utilise Kafka streaming. 

For a detailed description of the services' workflow see  
<!--- 
See doc/services.drawio for the source of this diagram
--->

> [Appendix I Palisade Flow Diagram](#appendix-i-palisade-flow-diagram).  

Reviewer: this link seems to be broken? where is it supposed to go?

Author: It is a link to the appendix at the end of this page where I have put the palisade flow diagram.

> For a more technical description of each of the services, follow the related links below.

Reviewer: ```suggestion
For a more technical description of each of the services, follow the links below.
```

<!--- 
See doc/services.drawio for the source of this diagram
--->
![Service Sequence diagram ](doc/services.png)

[Palisade Service](palisade-service/readme.md) takes client resource requests and starts the process.  
[User Service](user-service/readme.md) identifies the user in the context of the Palisade service.  
[Resource Service](resource-service/readme.md) identifies the resources that have been requested.  
[Policy Service](policy-service/readme.md) provides the policy that are to be applied to the data.  
[Attribute-Masking Service](attribute-masking-service/readme.md) creates a redacted version of the rescue.  
[Topic-Offset Service](topic-offset-service/readme.md) identifies the start and end of the resources for a request.  
[Filtered-Resource Service](filtered-resource-service/readme.md) readies the resources and later handles passing the data to the client.  
[Data Service](data-service/readme.md) initiates the process of passing the resources to the client.  

> [Audit Service](audit-service/readme.md) operates in the background providing an audit record of the process.  

Reviewer: Change this to be in alphabetical order

Author: done

# <img src="logos/logo.svg" width="180">

## A Tool for Complex and Scalable Data Access Policy Enforcement

## Overview

From the client’s perspective, they submit a request to examine data and receive a reference for this data. 
This reference can then be used to view the data after it has been retrieved and possibly redacted or filtered based on the context of the query and the permissions of the user making the request.

> Under the hood, the request involves sending a message that is processed sequentially by a set of the micro-services starting first with Palisade Service.

Reviewer 2: ```suggestion
Under the hood, the request involves sending a message that is processed sequentially by a set of micro-services starting first with Palisade Service.
```

The Palisade Service will perform two tasks.
First it will return a reference, a token back to the client that will uniquely identify this request.
This token is used throughout the application to tie every aspect of the process to this unique identifier and is later used by the client to retrieve the resources available for this request. 
The second task for the Palisade Service is to initiate the processing of the request by forwarding it onto the next service in the sequence (in this case the User Service).
The service will perform its required task and then forward the message onto the next service with the end goal of producing resources ready to be viewed by the client.
This sequence of steps is completed with the Filtered-Resource Service prepared to provide the processed data.

> In the subsequent request by the client to the Data Service, the service returns, one-by-one upon request over a websocket, metadata for a single resource (e.g. a file) they are authorised to view in some capacity.

Reviewer 2: ```suggestion
In the subsequent request by the client to the Filtered-Resource Service, the service returns, one-by-one upon request over a websocket, metadata for a single resource (e.g. a file) they are authorised to view in some capacity.
These resources can then be read from the Data Service.
```

The following diagram shows at a high level the services and their relative relationships with each other. 
Communications between the services and their related support servers utilise Kafka streaming. 

For a detailed description of the services' workflow see [Appendix I - Palisade Flow Diagram](#appendix-i).  
For a more technical description of each of the services, follow the links below.

<!--- 
See doc/services.drawio for the source of this diagram
--->
![Service Sequence diagram](doc/services.png)

[Palisade Service](palisade-service/README.md) takes client resource requests and starts the process.  

> [User Service](user-service/README.md) identifies the user in the context of the Palisade service.  

Reviewer: Although worded not quite right, something like below is more descriptive of the User Service
```
[User Service](user-service/README.md) ensures that the user is valid by checking they exist in a user provider technology.
```

Author: Went with the first suggestion for user service

[Resource Service](resource-service/README.md) identifies the resources that have been requested.  

> [Policy Service](policy-service/README.md) provides the policy that are to be applied to the data.  

Reviewer: ```suggestion
[Policy Service](policy-service/README.md) checks that policies exist for each resource in a request.  
```

> [Attribute-Masking Service](attribute-masking-service/README.md) creates a redacted version of the rescue.  

Reviewer: ```suggestion
[Attribute-Masking Service](attribute-masking-service/README.md) applies policies from the Policy Service against the resources in the request.
```

[Topic-Offset Service](topic-offset-service/README.md) identifies the start and end of the resources for a request.  
[Filtered-Resource Service](filtered-resource-service/README.md) readies the resources and later handles passing the data to the client.  

> [Data Service](data-service/README.md) initiates the process of passing the resources to the client.  

Reviewer: ```suggestion
[Data Service](data-service/README.md) retrieves resources from the relevant backing stores. 
```

> [Audit Service](audit-service/README.md) operates in the background providing an audit record of the process.    

Reviewer: ```suggestion
[Audit Service](audit-service/README.md) operates in the background providing an audit log for each request.    
```
